### PR TITLE
avoids setting jax tracer as lazy property attribute 

### PR DESCRIPTION
**Pull request by sbidari**
Add a conditional if `numpyro.util.not_jax_tracer(value)` before setting `lazy_property` as an attribute of `TruncatedDistribution`

This resolves errors described in #1836 and https://github.com/CDCgov/multisignal-epi-inference/issues/282

tests added:
check valid sampling of `TruncatedDistribution` in parallel (failed previously as described in #1836 )
check predictive methods (prior predictive and inference) can be run multiple times on the same model built using `TruncatedDistribution` (error encountered here https://github.com/CDCgov/multisignal-epi-inference/issues/282)

### CODE REVIEW

File: test/test_distributions.py
Comment (dylanhmorris) [inline test/test_distributions.py:3471]:
```suggestion
    n_prior_samples = 10
```

File: test/test_distributions.py
Comment (dylanhmorris) [inline test/test_distributions.py:3474]:
```suggestion
```

File: test/test_distributions.py
Comment (dylanhmorris) [inline test/test_distributions.py:3489]:
```suggestion
```

File: test/test_distributions.py
Comment (dylanhmorris) [inline test/test_distributions.py:3480]: Is one call to `prior_predictive()` sufficient to reproduce the error without the patch, or are two calls needed, as the reproducible example from https://github.com/CDCgov/multisignal-epi-inference/issues/282?

**Review by dylanhmorris — COMMENTED** [review]
Thanks for this, @sbidari. I wonder whether it might be worth annotating the test functions (via their docstrings) to explain why they should run without error if things are working and why/how they fail prior without the patch in this PR

File: test/test_distributions.py
Comment (sbidari) [reply test/test_distributions.py:3480]: Thanks @dylanhmorris. The second call is initiated when running the inference. 

But your comments made me realize I don't need two separate tests; the second test can check if the chains run in parallel and two calls to predictive method (prior predictive and inference run here)

**Review by fehiepsi — APPROVED** [review]
LGTM pending reviews from @dylanhmorris. Thanks @sbidari!

File: test/test_distributions.py
Comment (sbidari) [reply test/test_distributions.py:3458]: closing this as resolved in https://github.com/pyro-ppl/numpyro/pull/1843/commits/82dbc935323bb352e1fc8c1d2e58cabaef523108

**Comment by sbidari** [comment]
Thanks @fehiepsi!

@dylanhmorris I added docstring with links to where issues are described and combined the two test functions.

File: test/test_distributions.py
Comment (dylanhmorris) [inline test/test_distributions.py:3438]:
```suggestion
def test_no_tracer_leakage_in_truncated_distribution(my_dist):
```

**Review by dylanhmorris — APPROVED** [review]
Thanks, @sbidari. One suggestion you can take or leave on naming the test (since it should pass if there _isn't_ tracer leakage). @fehiepsi: this LGTM.

**Comment by sbidari** [comment]
I am not sure why the test is failing in CI with `UserWarning: There are not enough devices to run parallel chains: expected 2 but got 1.` despite using `numpyro.set_host_device_count(2)` at the beginning. Any thoughts @fehiepsi?

File: test/test_distributions.py
Comment (fehiepsi) [inline test/test_distributions.py:53]: It is better to avoid adding inference tests in this file. Maybe simple tests like are enough?
```
def test_jitted_lazy_property():
    d = dist.TruncatedNormal(...)
    jax.vmap(d.log_prob)(jnp.ones(3))
```

Comment (dylanhmorris) [reply test/test_distributions.py:53]: Somewhat related: since the key change is to the `distributions.util` module, I might suggest putting this test in 
`test_distributions_util.py` rather than `test_distributions.py`. Do you agree, @fehiepsi?

Comment (dylanhmorris) [reply test/test_distributions.py:53]: I like this idea, but at least for me the specific suggested test passes even with the existing `lazy_property` implementation (I have confirmed that the MRE from #1836 does indeed error).

**Comment by dylanhmorris** [comment]
Noting here that when this lands it will also provide a more general fix for #1651

File: test/test_distributions.py
Comment (fehiepsi) [reply test/test_distributions.py:53]: maybe calling vmap 2 times? it is surprised to me that jax does not catch this

Comment (fehiepsi) [reply test/test_distributions.py:53]: re distributiomz_util: that works :)

Comment (dylanhmorris) [reply test/test_distributions.py:53]: Have proposed a pair of tests in https://github.com/sbidari/numpyro/pull/1 and confirmed they fail with the latest `numpyro` on PyPI but pass with @sbidari's patch.

Comment (sbidari) [reply test/test_distributions.py]: Thanks @dylanhmorris for rewriting the tests! Ready for review @fehiepsi

**Review by fehiepsi — APPROVED** [review]
Thanks @sbidari!

**Comment by damonbayer** [comment]
@fehiepsi Is there a release planned for this fix? We would appreciate having it for [cdcgov/multisignal-epi-inference](https://github.com/cdcgov/multisignal-epi-inference/) as soon as possible.

**Comment by fehiepsi** [comment]
I can make a patch release in upcoming days. In the mean time, you can patch the utility.

**Comment by damonbayer** [comment]
@fehiepsi Do you have a date in mind?

**Pull request merged**
